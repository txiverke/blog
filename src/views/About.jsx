// @flow

import React from 'react'
import Helmet from 'react-helmet'
import { connect } from 'react-redux'
import ReactMessages from 'react-messages'

import Loader from '../components/Loader'
import ButtonBack from '../components/ButtonBack'
import { loadUserData } from '../actions/userActionCreators'
import avatar from '../assets/imgs/about.jpg'
import config from '../config'
import { getDictionary } from '../utils/dictionary'
// $FlowFixMe
import pdf from '../assets/files/XavierVilaAlbiol_CV_en.pdf'

class About extends React.Component {
  state = {
    txtMessage: '',
    next: false,
    DIC: {},
    sent: false
  }
  
  props: {
    user: Item,
    dispatch: Function,
    location: Object,
    language: Object
  }

  componentDidMount() {
    const { dispatch, user, language } = this.props

    this.setState({ DIC: getDictionary(language.current)})
    
    if (Object.keys(user.data).length === 0) {
      dispatch(loadUserData(config.api.profileId))
    }
  }

  componentWillReceiveProps(nextProps: Object) {
    const old = this.props.language.current
    const current = nextProps.language.current
    if (old !== current) {
      this.setState({ DIC: getDictionary(current)})
    }
  }

  renderMsg = () => {
    const { DIC } = this.state

    this.setState({
      txtMessage: DIC.RESUME,
      next: true,
      sent: true
    })

    setTimeout(() => this.setState({ next: false}))
  }

  render() {
    const { completed, data, message, error } = this.props.user
    const lang = this.props.language.current
    const { txtMessage, next, DIC, sent } = this.state
    const bio = lang === 'eng' ? data.bio : data.bio_cat
    const job = lang === 'eng' ? data.job : data.job_cat
    const btnMessage = sent ? DIC.DOWNLOADED_RESUME: DIC.DOWNLOAD_RESUME 

    if (completed) {
      return (
        <section className="app-view app-view-centered">
          <Helmet 
            title={DIC['ABOUT-ME']} 
            meta={[
              { name:"description", content: `${bio}` },
              { property: "og:title", content: `${data.firstname} ${data.lastname} - ${job}` }
            ]}
          />
          <ReactMessages message={txtMessage} next={next} error={error} icon={'thumbs-up'} />
          <ButtonBack label={DIC.BACK} />
          <article className="app-content-section">
            <figure className="app-content-avatar">
              <img 
                className="roundedImg" 
                src={avatar} 
                alt={`A pic of: ${data.firstname}`} 
            />
            </figure>
            <h1 className="tit-section-view">{data.firstname} {data.lastname}</h1>
            <h2 className="subtit-section-view">{job}</h2>
            <p className="txt txt-center">{bio}</p>
            <a
              href={pdf}
              download="resume.pdf" 
              className="btn btn-centered mt10"
              onClick={this.renderMsg}
            >
              {!sent && <span className="icon-cloud-download mr5"></span>}
              {sent && <span className="icon-like mr5"></span>}
              {btnMessage}
            </a>
          </article>
        </section>
      )
    }

    return (
      <div className="app-view app-view-centered">
        <Loader msg={message} />
      </div>
    )
  }
}
  
const mapStateToProps = state => ({ 
  user: state.user,
  language: state.language 
})

export const Unwrapped = About
export default connect(mapStateToProps)(About)