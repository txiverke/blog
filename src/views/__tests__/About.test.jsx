// @flow

import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { Unwrapped as UnwrappedAbout } from '../About'
import { shallow } from 'enzyme'

const properties = {
  user: { 
    completed: true,
    data: {
      bio: 'test bio',
      firstname: 'test firstname',
      lastname: 'test lastname',
      job: 'test job',
      bio: 'test bio'
    },
    error: false,
    message: 'Just a fake message'
  },
  dispatch: jest.fn(),
  location: {},
  language: { current: 'eng'}
}

describe('<About />', () => {
  it('should render itself', () => {
    const component = shallow(<UnwrappedAbout {...properties} />)
    expect(component).toMatchSnapshot()
  })

  it('should render loader if data is not completed', () => {
    properties.user.completed = false
    const component = shallow(<UnwrappedAbout {...properties} />)    
    expect(component.find('Loader').length).toEqual(1)
  })

  it('should call renderMsg and update the btn state', () => {
    properties.user.completed = true
    const component = shallow(<UnwrappedAbout {...properties} />)
    expect(component.find('.icon-cloud-download').length).toEqual(1)
    expect(component.find('.icon-like').length).toEqual(0)
    expect(component.find('a.btn-centered').text()).toEqual('Download Résumé')
    component.find('a.btn-centered').simulate('click')
    expect(component.find('.icon-cloud-download').length).toEqual(0)
    expect(component.find('.icon-like').length).toEqual(1)
    expect(component.find('a.btn-centered').text()).toEqual('Résumé downloaded')
  })
})