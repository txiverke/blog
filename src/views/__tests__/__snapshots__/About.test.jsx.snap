// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<About /> should render itself 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <About
    dispatch={[Function]}
    language={
        Object {
            "current": "eng",
          }
    }
    location={Object {}}
    user={
        Object {
            "completed": true,
            "data": Object {
              "bio": "test bio",
              "firstname": "test firstname",
              "job": "test job",
              "lastname": "test lastname",
            },
            "error": false,
            "message": "Just a fake message",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <HelmetWrapper
          defer={true}
          encodeSpecialCharacters={true}
          meta={
                    Array [
                              Object {
                                "content": "test bio",
                                "name": "description",
                              },
                              Object {
                                "content": "test firstname test lastname - test job",
                                "property": "og:title",
                              },
                            ]
          }
          title="ABOUT ME"
/>,
        <M
          duration={2000}
          error={false}
          icon="thumbs-up"
          message=""
          next={false}
/>,
        <ButtonBack
          label="Back"
          route="/"
/>,
        <article
          className="app-content-section"
>
          <figure
                    className="app-content-avatar"
          >
                    <img
                              alt="A pic of: test firstname"
                              className="roundedImg"
                              src="about.jpg"
                    />
          </figure>
          <h1
                    className="tit-section-view"
          >
                    test firstname
                     
                    test lastname
          </h1>
          <h2
                    className="subtit-section-view"
          >
                    test job
          </h2>
          <p
                    className="txt txt-center"
          >
                    test bio
          </p>
          <a
                    className="btn btn-centered mt10"
                    download="resume.pdf"
                    href="XavierVilaAlbiol_CV_en.pdf"
                    onClick={[Function]}
          >
                    <span
                              className="icon-cloud-download mr5"
                    />
                    Download Résumé
          </a>
</article>,
      ],
      "className": "app-view app-view-centered",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "defer": true,
          "encodeSpecialCharacters": true,
          "meta": Array [
            Object {
              "content": "test bio",
              "name": "description",
            },
            Object {
              "content": "test firstname test lastname - test job",
              "property": "og:title",
            },
          ],
          "title": "ABOUT ME",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "duration": 2000,
          "error": false,
          "icon": "thumbs-up",
          "message": "",
          "next": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "label": "Back",
          "route": "/",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <figure
              className="app-content-avatar"
>
              <img
                            alt="A pic of: test firstname"
                            className="roundedImg"
                            src="about.jpg"
              />
</figure>,
            <h1
              className="tit-section-view"
>
              test firstname
               
              test lastname
</h1>,
            <h2
              className="subtit-section-view"
>
              test job
</h2>,
            <p
              className="txt txt-center"
>
              test bio
</p>,
            <a
              className="btn btn-centered mt10"
              download="resume.pdf"
              href="XavierVilaAlbiol_CV_en.pdf"
              onClick={[Function]}
>
              <span
                            className="icon-cloud-download mr5"
              />
              Download Résumé
</a>,
          ],
          "className": "app-content-section",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <img
                alt="A pic of: test firstname"
                className="roundedImg"
                src="about.jpg"
/>,
              "className": "app-content-avatar",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "A pic of: test firstname",
                "className": "roundedImg",
                "src": "about.jpg",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            "type": "figure",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "test firstname",
                " ",
                "test lastname",
              ],
              "className": "tit-section-view",
            },
            "ref": null,
            "rendered": Array [
              "test firstname",
              " ",
              "test lastname",
            ],
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "test job",
              "className": "subtit-section-view",
            },
            "ref": null,
            "rendered": "test job",
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "test bio",
              "className": "txt txt-center",
            },
            "ref": null,
            "rendered": "test bio",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <span
                  className="icon-cloud-download mr5"
/>,
                false,
                "Download Résumé",
              ],
              "className": "btn btn-centered mt10",
              "download": "resume.pdf",
              "href": "XavierVilaAlbiol_CV_en.pdf",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "icon-cloud-download mr5",
                },
                "ref": null,
                "rendered": null,
                "type": "span",
              },
              false,
              "Download Résumé",
            ],
            "type": "a",
          },
        ],
        "type": "article",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <HelmetWrapper
            defer={true}
            encodeSpecialCharacters={true}
            meta={
                        Array [
                                    Object {
                                      "content": "test bio",
                                      "name": "description",
                                    },
                                    Object {
                                      "content": "test firstname test lastname - test job",
                                      "property": "og:title",
                                    },
                                  ]
            }
            title="ABOUT ME"
/>,
          <M
            duration={2000}
            error={false}
            icon="thumbs-up"
            message=""
            next={false}
/>,
          <ButtonBack
            label="Back"
            route="/"
/>,
          <article
            className="app-content-section"
>
            <figure
                        className="app-content-avatar"
            >
                        <img
                                    alt="A pic of: test firstname"
                                    className="roundedImg"
                                    src="about.jpg"
                        />
            </figure>
            <h1
                        className="tit-section-view"
            >
                        test firstname
                         
                        test lastname
            </h1>
            <h2
                        className="subtit-section-view"
            >
                        test job
            </h2>
            <p
                        className="txt txt-center"
            >
                        test bio
            </p>
            <a
                        className="btn btn-centered mt10"
                        download="resume.pdf"
                        href="XavierVilaAlbiol_CV_en.pdf"
                        onClick={[Function]}
            >
                        <span
                                    className="icon-cloud-download mr5"
                        />
                        Download Résumé
            </a>
</article>,
        ],
        "className": "app-view app-view-centered",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "defer": true,
            "encodeSpecialCharacters": true,
            "meta": Array [
              Object {
                "content": "test bio",
                "name": "description",
              },
              Object {
                "content": "test firstname test lastname - test job",
                "property": "og:title",
              },
            ],
            "title": "ABOUT ME",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "duration": 2000,
            "error": false,
            "icon": "thumbs-up",
            "message": "",
            "next": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "label": "Back",
            "route": "/",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <figure
                className="app-content-avatar"
>
                <img
                                alt="A pic of: test firstname"
                                className="roundedImg"
                                src="about.jpg"
                />
</figure>,
              <h1
                className="tit-section-view"
>
                test firstname
                 
                test lastname
</h1>,
              <h2
                className="subtit-section-view"
>
                test job
</h2>,
              <p
                className="txt txt-center"
>
                test bio
</p>,
              <a
                className="btn btn-centered mt10"
                download="resume.pdf"
                href="XavierVilaAlbiol_CV_en.pdf"
                onClick={[Function]}
>
                <span
                                className="icon-cloud-download mr5"
                />
                Download Résumé
</a>,
            ],
            "className": "app-content-section",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <img
                  alt="A pic of: test firstname"
                  className="roundedImg"
                  src="about.jpg"
/>,
                "className": "app-content-avatar",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "A pic of: test firstname",
                  "className": "roundedImg",
                  "src": "about.jpg",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              "type": "figure",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "test firstname",
                  " ",
                  "test lastname",
                ],
                "className": "tit-section-view",
              },
              "ref": null,
              "rendered": Array [
                "test firstname",
                " ",
                "test lastname",
              ],
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "test job",
                "className": "subtit-section-view",
              },
              "ref": null,
              "rendered": "test job",
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "test bio",
                "className": "txt txt-center",
              },
              "ref": null,
              "rendered": "test bio",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <span
                    className="icon-cloud-download mr5"
/>,
                  false,
                  "Download Résumé",
                ],
                "className": "btn btn-centered mt10",
                "download": "resume.pdf",
                "href": "XavierVilaAlbiol_CV_en.pdf",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "icon-cloud-download mr5",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "span",
                },
                false,
                "Download Résumé",
              ],
              "type": "a",
            },
          ],
          "type": "article",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
